packages/wallet-sdk/src/sign/scw/utils/createSmartAccount.ts

```diff
@@ -0,0 +1,830 @@
+  encodePacked,
+  type Hash,
+  hashMessage,
+  hashTypedData,
+  type Hex,
+  type LocalAccount,
+  type OneOf,
+  parseSignature,
+  type Prettify,
+  size,
+  stringToHex,
+  type TypedDataDefinition,
+} from 'viem';
+import {
+  entryPoint06Abi,
+  entryPoint06Address,
+  getUserOperationHash,
+  type SmartAccount,
+  type SmartAccountImplementation,
+  toSmartAccount,
+  type UserOperation,
+  type WebAuthnAccount,
+} from 'viem/account-abstraction';
+
+export type CreateSmartAccountParameters = {
+  address: Address;
+  client: Client;
+  factoryData: Hex | undefined;
+  accountIndex: number;
+  account: OneOf<LocalAccount | WebAuthnAccount>;
+};
+
+export type CreateSmartAccountReturnType = Prettify<
+  SmartAccount<CoinbaseSmartAccountImplementation>
+>;
+
+export type CoinbaseSmartAccountImplementation = Assign<
+  SmartAccountImplementation<
+    typeof entryPoint06Abi,
+    '0.6',
+    { abi: typeof abi; factory: { abi: typeof factoryAbi; address: Address } }
+  >,
+  {
+    decodeCalls: NonNullable<SmartAccountImplementation['decodeCalls']>;
+    sign: NonNullable<SmartAccountImplementation['sign']>;
+  }
+>;
+
+/**
+ * @description Create a Coinbase Smart Account.
+ *
+ * @param parameters - {@link CreateSmartAccountParameters}
+ * @returns Coinbase Smart Account. {@link CreateSmartAccountReturnType}
+ *
+ * @example
+ * import { createSmartAccount } from 'viem/account-abstraction'
+ * import { privateKeyToAccount } from 'viem/accounts'
+ * import { client } from './client.js'
+ *
+ * const account = createSmartAccount({
+ *   client,
+ *   account: privateKeyToAccount('0x...'),
+ *   accountIndex: 0,
+ *   address: '0x...',
+ *   factoryData: '0x...',
+ * })
+ */
+export async function createSmartAccount(
+  parameters: CreateSmartAccountParameters
+): Promise<CreateSmartAccountReturnType> {
+  const { account, accountIndex, address, client, factoryData } = parameters;
+  const entryPoint = {
+    abi: entryPoint06Abi,
+    address: entryPoint06Address,
+    version: '0.6',
+  } as const;
+  const factory = {
+    abi: factoryAbi,
+    address: '0x0ba5ed0c6aa8c49038f819e587e2633c4a9f428a',
+  } as const;
+
+  const owner = (() => {
+    const owner = account;
+    if (typeof owner === 'string') return { address: owner, type: 'address' } as const;
+    return owner;
+  })();
+
+  return toSmartAccount({
+    client,
+    entryPoint,
+    extend: { abi, factory },
+    async decodeCalls(data) {
+      const result = decodeFunctionData({
+        abi,
+        data,
+      });
+      if (result.functionName === 'execute')
+        return [{ to: result.args[0], value: result.args[1], data: result.args[2] }];
+      if (result.functionName === 'executeBatch')
+        return result.args[0].map((arg) => ({
+          to: arg.target,
+          value: arg.value,
+          data: arg.data,
+        }));
+      throw new BaseError(`unable to decode calls for "${result.functionName}"`);
+    },
+
+    async encodeCalls(calls) {
```

Comment:
do we need to handle the 0 length array case?

Reply:
I think we're ok. I've been testing with zero length arrays currently and everything is working